src/client/texture_paths.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+Minetest
+Copyright (C) 2010-2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "texture_paths.h"
+
+#include "util/container.h"
+#include "settings.h"
+#include "filesys.h"
+#include "porting.h"
+
+// A cache from texture name to texture path
+static MutexedMap<std::string, std::string> g_texturename_to_path_cache;
+
+void clearTextureNameCache()
+{
+	g_texturename_to_path_cache.clear();
+}
+
+// Find out the full path of an image by trying different filename extensions.
+// If failed, return "".
+std::string getImagePath(std::string_view path)
+{
+	// A NULL-ended list of possible image extensions
+	const char *extensions[] = {".png", ".jpg", ".bmp", ".tga", NULL};
+	// Remove present extension
+	std::string_view stripped_path = removeStringEnd(path, extensions);
+
+	// Check paths until something is found to exist
+	const char **ext = extensions;
+	do {
+		std::string extended_path{stripped_path};
+		extended_path.append(*ext);
+		if (fs::PathExists(extended_path))
+			return extended_path;
+	}
+	while(*(++ext) != NULL);
```

Comment:
a `for (auto &ext : extensions)` would be more C++-y, just be careful of the null value

Reply:
Ok, I modernized it.
Note that it now uses a vector instead of a C style array, since as far as I know, C++17 does not have `ranges::subrange` yet and I don't think you can limit the size of a C++ array without using additional memory.